Support Nonlinear.NODE_LOGIC and Nonlinear.NODE_COMPARISON

Hi @odow,

I've tried this backend with my optimal transport network models at https://github.com/SebKrantz/OptimalTransportNetworks.jl/tree/main/src/models, and it seems to work for none of them. For example, using [Example 1](https://github.com/SebKrantz/OptimalTransportNetworks.jl/blob/main/examples/example01.jl) which calls [this model](https://github.com/SebKrantz/OptimalTransportNetworks.jl/blob/main/src/models/model_mobility.jl):

```julia
using OptimalTransportNetworks

param = init_parameters(labor_mobility = true, K = 10, gamma = 1, beta = 1, verbose = true,
                        N = 1, kappa_tol = 1e-5, cross_good_congestion = false, nu = 1)

param, graph = create_graph(param, 11, 11, type = "map") # create a map network of 11x11 nodes located in [0,10]x[0,10]

# Customize graph
param[:Zjn] = fill(0.1, param[:J], param[:N]) # set most places to low productivity
Ni = find_node(graph, 6, 6) # Find index of the central node at (6,6)
param[:Zjn][Ni, :] .= 1 # central node more productive

# Optimizer options 
import MathOptInterface as MOI
import MathOptSymbolicAD
param[:optimizer_attr] = Dict(:linear_solver => "ma57") 
param[:model_attr] = Dict(:backend => (MOI.AutomaticDifferentiationBackend(), MathOptSymbolicAD.DefaultBackend()))

# RESOLUTION
@time results = optimal_network(param, graph) # solve the network

# Plot the network with the optimal transportation plan

plot_graph(graph, results[:Ijk], node_sizes = results[:Cj])
```

Yields:
```
A JuMP Model
Maximization problem with:
Variables: 1204
Objective function type: VariableRef
`NonlinearExpr`-in-`MathOptInterface.LessThan{Float64}`: 242 constraints
`AffExpr`-in-`MathOptInterface.Interval{Float64}`: 122 constraints
`VariableRef`-in-`MathOptInterface.GreaterThan{Float64}`: 363 constraints
`VariableRef`-in-`MathOptInterface.LessThan{Float64}`: 121 constraints
`VariableRef`-in-`MathOptInterface.Parameter{Float64}`: 420 constraints
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: Ipopt
Names registered in the model: Cjn, Lj, Ljn, Pjn, Qin, Yjn, kappa_ex, u

ERROR: MathOptInterface.UnsupportedAttribute{MathOptInterface.AutomaticDifferentiationBackend}: Attribute MathOptInterface.AutomaticDifferentiationBackend() is not supported by the model.
Stacktrace:
 [1] throw_set_error_fallback(model::Ipopt.Optimizer, attr::MathOptInterface.AutomaticDifferentiationBackend, value::MathOptSymbolicAD.DefaultBackend; error_if_supported::MathOptInterface.SetAttributeNotAllowed{MathOptInterface.AutomaticDifferentiationBackend})
   @ MathOptInterface ~/.julia/packages/MathOptInterface/mz9FK/src/attributes.jl:586
 [2] throw_set_error_fallback(model::Ipopt.Optimizer, attr::MathOptInterface.AutomaticDifferentiationBackend, value::MathOptSymbolicAD.DefaultBackend)
   @ MathOptInterface ~/.julia/packages/MathOptInterface/mz9FK/src/attributes.jl:577
 [3] set(model::Ipopt.Optimizer, attr::MathOptInterface.AutomaticDifferentiationBackend, args::MathOptSymbolicAD.DefaultBackend)
   @ MathOptInterface ~/.julia/packages/MathOptInterface/mz9FK/src/attributes.jl:550
 [4] set(b::MathOptInterface.Bridges.LazyBridgeOptimizer{Ipopt.Optimizer}, attr::MathOptInterface.AutomaticDifferentiationBackend, value::MathOptSymbolicAD.DefaultBackend)
   @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/mz9FK/src/Bridges/bridge_optimizer.jl:955
 [5] set(model::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{Ipopt.Optimizer}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}, attr::MathOptInterface.AutomaticDifferentiationBackend, value::MathOptSymbolicAD.DefaultBackend)
   @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/mz9FK/src/Utilities/cachingoptimizer.jl:1059
 [6] set(m::Model, attr::MathOptInterface.AutomaticDifferentiationBackend, value::MathOptSymbolicAD.DefaultBackend)
   @ JuMP ~/.julia/packages/JuMP/as6Ji/src/optimizer_interface.jl:794
 [7] set_attribute(model::Model, attr::MathOptInterface.AutomaticDifferentiationBackend, value::MathOptSymbolicAD.DefaultBackend)
   @ JuMP ~/.julia/packages/JuMP/as6Ji/src/optimizer_interface.jl:1023
 [8] top-level scope
   @ ~/Documents/Julia/OptimalTransportNetworks.jl/examples/example01.jl:59
```
This does not seem to be dependent on how the parameters are passed to the optimizer, for example I can 

```julia
model, recover_allocation = optimal_network(param, graph, verbose = true, return_model = true)
set_attribute(
    model,
    MOI.AutomaticDifferentiationBackend(),
    MathOptSymbolicAD.DefaultBackend(),
)
```
Which yields the same. I note that this works with Matlab's symbolic differentiation engine [ADiGator](https://github.com/matt-weinstein/adigator), so I guess the problem lies in certain parts of the model specification not supported by the symbolic backend rather than the model not being symbolically differentiable?

Thanks @odow, indeed, my version of Ipopt was outdated, but even after updating I get

```
ERROR: AssertionError: node.type == MOI.Nonlinear.NODE_VALUE
Stacktrace:
  [1] _to_expr(data::MathOptInterface.Nonlinear.Model, expr::MathOptInterface.Nonlinear.Expression, variable_order::Dict{Int64, Int64}, subexpressions::Vector{Any})
    @ MathOptSymbolicAD ~/.julia/packages/MathOptSymbolicAD/A5rTJ/src/nonlinear_oracle.jl:521
  [2] (::MathOptSymbolicAD.var"#16#18"{MathOptInterface.Nonlinear.Model, Vector{Any}, Dict{Int64, Int64}})(c::MathOptInterface.Nonlinear.Constraint)
    @ MathOptSymbolicAD ~/.julia/packages/MathOptSymbolicAD/A5rTJ/src/nonlinear_oracle.jl:549
  [3] iterate
    @ ./generator.jl:47 [inlined]
  [4] collect_to!
    @ ./array.jl:845 [inlined]
  [5] collect_to_with_first!
    @ ./array.jl:823 [inlined]
  [6] collect(itr::Base.Generator{Base.ValueIterator{OrderedCollections.OrderedDict{MathOptInterface.Nonlinear.ConstraintIndex, MathOptInterface.Nonlinear.Constraint}}, MathOptSymbolicAD.var"#16#18"{MathOptInterface.Nonlinear.Model, Vector{Any}, Dict{Int64, Int64}}})
    @ Base ./array.jl:797
  [7] map
    @ ./abstractarray.jl:2961 [inlined]
  [8] MathOptInterface.Nonlinear.Evaluator(model::MathOptInterface.Nonlinear.Model, backend::MathOptSymbolicAD.DefaultBackend, ordered_variables::Vector{MathOptInterface.VariableIndex})
    @ MathOptSymbolicAD ~/.julia/packages/MathOptSymbolicAD/A5rTJ/src/nonlinear_oracle.jl:548
  [9] _setup_model(model::Ipopt.Optimizer)
    @ Ipopt ~/.julia/packages/Ipopt/YDBAD/src/MOI_wrapper.jl:789
 [10] optimize!(model::Ipopt.Optimizer)
    @ Ipopt ~/.julia/packages/Ipopt/YDBAD/src/MOI_wrapper.jl:900
 [11] optimize!
    @ ~/.julia/packages/MathOptInterface/mz9FK/src/Bridges/bridge_optimizer.jl:380 [inlined]
 [12] optimize!
    @ ~/.julia/packages/MathOptInterface/mz9FK/src/MathOptInterface.jl:85 [inlined]
 [13] optimize!(m::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{Ipopt.Optimizer}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}})
    @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/mz9FK/src/Utilities/cachingoptimizer.jl:316
 [14] optimize!(model::JuMP.Model; ignore_optimize_hook::Bool, _differentiation_backend::MathOptInterface.Nonlinear.SparseReverseMode, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JuMP ~/.julia/packages/JuMP/as6Ji/src/optimizer_interface.jl:457
 [15] optimize!(model::JuMP.Model)
    @ JuMP ~/.julia/packages/JuMP/as6Ji/src/optimizer_interface.jl:409
 [16] optimal_network(param::Dict{Any, Any}, graph::Dict{Symbol, Any}; I0::Nothing, Il::Nothing, Iu::Nothing, verbose::Bool, return_model::Int64)
    @ OptimalTransportNetworks ~/Documents/Julia/OptimalTransportNetworks.jl/src/main/optimal_network.jl:156
 [17] top-level scope
    @ ./timing.jl:262 [inlined]
 [18] top-level scope
    @ ~/Documents/Julia/OptimalTransportNetworks.jl/examples/example01.jl:0
```
For both the fixed and mobile model. I thus presume this has to do with some part of the model specification. Perhaps the parameters?

--- a/src/nonlinear_oracle.jl
+++ b/src/nonlinear_oracle.jl
@@ -173,6 +173,9 @@ Parse `expr` into `f` and return a `_Node`.
     This function gets called recursively.
 """
 function _Node(f::_Function, expr::Expr)
+    if Meta.isexpr(expr, :||) || Meta.isexpr(expr, :&&)
+        return _Node(f, expr.head, (_Node(f, arg) for arg in expr.args)...)
+    end
     @assert Meta.isexpr(expr, :call)
     # Performance optimization: most calls will be unary or binary
     # operators. Therefore, we can specialize an if-statement to handle the
@@ -219,6 +222,10 @@ function _expr_to_symbolics(model::MOI.Nonlinear.Model, expr::_Node, p, x)
         args = [_expr_to_symbolics(model, c, p, x) for c in expr.children]
         if hasproperty(Base, expr.operation)
             return getproperty(Base, expr.operation)(args...)
+        elseif expr.operation == :&&
+            return (&)(args...)
+        elseif expr.operation == :||
+            return (|)(args...)
         end
         # If the function isn't defined in Base, defer to the operator registry.
         # We don't do this for all functions, because MOI uses NaNMath, which
@@ -518,6 +525,10 @@ function _to_expr(
             MOI.VariableIndex(variable_order[node.index])
         elseif node.type == MOI.Nonlinear.NODE_PARAMETER
             data.parameters[node.index]
+        elseif node.type == MOI.Nonlinear.NODE_LOGIC
+            Expr(data.operators.logic_operators[node.index])
+        elseif node.type == MOI.Nonlinear.NODE_COMPARISON
+            Expr(:call, data.operators.comparison_operators[node.index])
         else
             @assert node.type == MOI.Nonlinear.NODE_VALUE
             expr.values[node.index]

--- a/test/runtests.jl
+++ b/test/runtests.jl
@@ -422,6 +422,9 @@ function test_constant_subexpressions_expr()
 end
 
 function test_logic_comparison_expr()
+    if VERSION < v"1.7"
+        return  # Symbolics doesn't support Base.ifelse in Julia v1.6
+    end
     model = Model(Ipopt.Optimizer)
     @variable(model, -1 <= x <= 1)
     @objective(model, Max, ifelse(-0.5 <= x && x <= 0.5, 1 - x^2, 0))

--- a/test/runtests.jl
+++ b/test/runtests.jl
@@ -421,6 +421,20 @@ function test_constant_subexpressions_expr()
     return
 end
 
+function test_logic_comparison_expr()
+    model = Model(Ipopt.Optimizer)
+    @variable(model, -1 <= x <= 1)
+    @objective(model, Max, ifelse(-0.5 <= x && x <= 0.5, 1 - x^2, 0))
+    set_attribute(
+        model,
+        MOI.AutomaticDifferentiationBackend(),
+        MathOptSymbolicAD.DefaultBackend(),
+    )
+    optimize!(model)
+    Test.@test termination_status(model) == LOCALLY_SOLVED
+    return
+end
+
 end  # module
 
 RunTests.runtests()

